tools/bin/prepare_release

```diff
@@ -0,0 +1,68 @@
+#!/usr/bin/env bash
+
+set -e
+
+case "$1" in
+minor|Minor ) BUMP="Minor" ;;
+patch|Patch ) BUMP="Patch" ;;
+* ) echo "Missing release type [minor|patch]"; exit 1;;
+esac
+MAGE_BUMP="version:bump$BUMP"
+
+export MAGEFILE_VERSBOSE=0
+
+MAGE=${MAGE:-tools/bin/mage}
+ORIGIN=${ORIGIN:-origin}
+
+MINOR_BRANCH=$($MAGE "$MAGE_BUMP" version:majorMinor)
+RELEASE_TAG=$($MAGE "$MAGE_BUMP" version:current)
+RELEASE_DATE=$(date +%Y-%m-%d)
+RELEASE_BRANCH=release/$RELEASE_TAG
+
+echo "Current Version: $($MAGE version:current)"
+echo "$BUMP bump branch $MINOR_BRANCH to $RELEASE_TAG"
+
+echo "Checking out $RELEASE_BRANCH"
+git checkout -b "$RELEASE_BRANCH"
+
+echo "Pulling latest changes for submodules"
+$MAGE git:pullSubmodules
+
+if [[ ! -z "$(git diff --name-only data)" ]]; then
+  echo "Committing changes for updated submodules"
+  git add data
+  git commit -m "data: Update external repositories"
+fi
+
+echo "Updating CHANGELOG.md"
+
+# Take care of the [3.x.y] section
+cat CHANGELOG.md | tr '\n' '\000' `# First change newlines to \x00 so that sed can treat whole file as a stream of text` | \
+    sed 's/## [A-Z][a-z]*\x00\x00#/\x01\x00#/g' `# Then replace empty sections with \x01, we mark them for deletion later` | \
+    # The reason why we mark for deletion and then delete later is because otherwise there are issues when two consecutive sections are empty
+    sed 's/## \[Unreleased\]/## \[Unreleased\]\x00\x00### Added\x00\x00### Changed\x00\x00### Deprecated\x00\x00### Removed\x00\x00### Fixed\x00\x00### Security\x00\x00'"## \\[${RELEASE_TAG:1}\\] - ${RELEASE_DATE}"'/' `# Rename the "Unreleased" section to "Tag - Date", and add a new, empty one` | \
+    tr '\000' '\n' `# Change \x00 back to newlines` | \
+    sed '/\x01/d' `# Remove lines with \x01 (previously marked for deletion)` | \
+    sed "s/\[unreleased\]: https:\/\/github.com\/$(echo "${GITHUB_REPOSITORY}" | sed 's/\//\\\//g')\/compare\/\(v3\.[0-9][0-9]*\.[0-9][0-9]*\)\.\.\.\(v3\.[0-9][0-9]*\)/[unreleased]: https:\/\/github.com\/$(echo "${GITHUB_REPOSITORY}" | sed 's/\//\\\//g')\/compare\/${RELEASE_TAG}...${MINOR_BRANCH}\n[${RELEASE_TAG}]: https:\/\/github.com\/$(echo "${GITHUB_REPOSITORY}" | sed 's/\//\\\//g')\/compare\/\1...${RELEASE_TAG}/" > CHANGELOG2.md # Add new github diff link
```

Comment:
There must be a better way to do this. Can't we script something in Go (Mage) or JS?

If there really isn't a better way, I'm going to have to take a better look at this later before I can approve this PR.

Reply:
> There must be a better way to do this

Well, the "correct" way would be to write a parser, but then we would have lots of code to perform a task that can be done by a few calls to `sed`.

Reply:
FYI: When I was working on this myself I wrote a thing in Go. The (opinionated) formatting of the rendering library makes everything look slightly differently than we have now, and I thought forking the library was a bit overkill, but maybe interesting to consider:

```go
package main

import (
	"bytes"
	"fmt"
	"log"
	"os"
	"time"

	"github.com/jsternberg/markdownfmt/markdown"
	"gopkg.in/russross/blackfriday.v2"
)

const (
	oldVersion = "3.17.0"
	newVersion = "3.17.1"
	branch     = "3.17"
)

const unreleasedSection = `## [Unreleased](%s)

### Added

### Changed

### Deprecated

### Removed

### Fixed

### Security
`

var unreleasedNodes = func() []*blackfriday.Node {
	var nodes []*blackfriday.Node
	unreleasedAST := newParser().Parse([]byte(fmt.Sprintf(unreleasedSection, compareLink(newVersion, branch))))
	unreleasedAST.Walk(func(node *blackfriday.Node, entering bool) blackfriday.WalkStatus {
		if !entering || node.Type == blackfriday.Document {
			return blackfriday.GoToNext
		}
		nodes = append(nodes, node)
		return blackfriday.SkipChildren
	})
	return nodes
}()

func newParser() *blackfriday.Markdown {
	const extensions = blackfriday.NoIntraEmphasis |
		blackfriday.Tables |
		blackfriday.FencedCode |
		blackfriday.Autolink |
		blackfriday.Strikethrough |
		blackfriday.SpaceHeadings |
		blackfriday.NoEmptyLineBeforeBlock

	return blackfriday.New(
		blackfriday.WithExtensions(extensions),
	)
}

func compareLink(oldVersion, newVersion string) string {
	return fmt.Sprintf("https://github.com/TheThingsNetwork/lorawan-stack/compare/v%s...v%s", oldVersion, newVersion)
}

func main() {
	input, err := os.ReadFile("CHANGELOG.md")
	if err != nil {
		log.Fatal(err)
	}

	var (
		changelogAST      = newParser().Parse(input)
		newReleaseHeading *blackfriday.Node
		toDelete          []*blackfriday.Node
	)

	changelogAST.Walk(func(node *blackfriday.Node, entering bool) blackfriday.WalkStatus {
		if !entering || node.Type != blackfriday.Heading {
			return blackfriday.GoToNext
		}
		switch node.HeadingData.Level {
		case 2: // H2: Version heading
			link := node.FirstChild.Next
			if link == nil || link.Type != blackfriday.Link {
				log.Fatalf("Version heading missing a link")
			}
			version := link.FirstChild
			if string(version.Literal) != "Unreleased" {
				return blackfriday.GoToNext
			}
			newReleaseHeading = node
		case 3: // H3: Added / Changed / Fixed / etc.
			var hasContents bool
			for n := node.Next; n != nil && n.Type != blackfriday.Heading; n = n.Next {
				hasContents = true
			}
			if !hasContents {
				toDelete = append(toDelete, node)
			}
		}
		return blackfriday.GoToNext
	})

	// Update Unreleased heading to new version:
	newReleaseHeading.FirstChild.Next.Destination = []byte(compareLink(oldVersion, newVersion))
	newReleaseHeading.FirstChild.Next.FirstChild.Literal = []byte(newVersion)
	newReleaseHeading.AppendChild(&blackfriday.Node{
		Type:    blackfriday.Text,
		Literal: []byte(fmt.Sprintf(" - %s", time.Now().Format("2006-01-02"))),
	})

	// Remove empty headings:
	for _, node := range toDelete {
		node.Unlink()
	}

	// Insert new Unreleased section:
	for _, node := range unreleasedNodes {
		newReleaseHeading.InsertBefore(node)
	}

	// Render back to Markdown:
	var buf bytes.Buffer
	renderer := markdown.NewRenderer(&markdown.Options{})
	renderer.RenderHeader(&buf, changelogAST)
	changelogAST.Walk(func(node *blackfriday.Node, entering bool) blackfriday.WalkStatus {
		return renderer.RenderNode(&buf, node, entering)
	})

	// Write to formatted file:
	err = os.WriteFile("CHANGELOG.fmt.md", buf.Bytes(), 0644)
	if err != nil {
		log.Fatal(err)
	}
}
```